AdminServer/appscale/admin/instance_manager/source_manager.py

```diff
@@ -37,6 +43,61 @@ def __init__(self, zk_client, thread_pool):
     self.zk_client = zk_client
     self.thread_pool = thread_pool
     self.source_futures = {}
+    self.projects_manager = None
+    self.fetched_revisions = None
+
+  def configure_automatic_fetch(self, project_manager):
+    """ Ensures that SourceManager watches zookeeper nodes describing
+    application archives and fetches it as soon as new archive
+    is available.
+
+    Args:
+      project_manager: an instance of GlobalProjectManager.
+    """
+    if self.projects_manager is not None:
+      logger.debug("Automatic fetch of new application archives "
+                   "was already configured")
+      return
+    self.projects_manager = project_manager
+    self.fetched_revisions = set()
+    self.zk_client.ChildrenWatch('/apps', self._update_apps_watch)
+    logger.debug("Configured automatic fetch of new application archives")
+
+  def _update_apps_watch(self, new_apps_list):
+    """ Schedules fetch of new sources if there are any available.
+
+    Args:
+      new_apps_list: a list of revisions with archive
+        available somewhere in deployment.
+    """
+    persistent_update_apps = retry_children_watch_coroutine(
+      '/apps', self._handle_apps_update
+    )
+    main_io_loop = IOLoop.current()
+    main_io_loop.add_callback(persistent_update_apps, new_apps_list)
+
+  @gen.coroutine
+  def _handle_apps_update(self, new_apps_list):
+    """ Fetches new available sources if there are any.
+
+    Args:
+      new_apps_list: a list of revisions with archive
+        available somewhere in deployment.
+    """
+    for revision_key in new_apps_list:
+      if revision_key in self.fetched_revisions:
+        continue
+      revision_key_parts = revision_key.split(VERSION_PATH_SEPARATOR)
+      project_id = revision_key_parts[0]
+      service_id = revision_key_parts[1]
+      version_id = revision_key_parts[2]
+
+      service_manager = self.projects_manager[project_id][service_id]
+      version_details = service_manager[version_id].version_details
+      runtime = version_details['runtime']
+      source_archive = version_details['deployment']['zip']['sourceUrl']
+      yield self.ensure_source(revision_key, source_archive, runtime)
```

Comment:
I would recommend reducing the new_apps_list to a set of versions. The sourceUrl from the version node will not match an older revision from /apps, and it would not make sense to fetch an older revision anyway (if, for example, the AppManager gets restarted and loses track of its fetched versions).